internal/checker/checker.go

```diff
@@ -21174,14 +21176,64 @@ func (c *Checker) instantiateTypeWithAlias(t *Type, m *TypeMapper, alias *TypeAl
 		c.error(c.currentNode, diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite)
 		return c.errorType
 	}
+	index := c.findActiveMapper(m)
+	if index == -1 {
+		c.pushActiveMapper(m)
+	}
+	var b KeyBuilder
+	b.WriteType(t)
+	b.WriteAlias(alias)
+	key := b.String()
+	cache := c.activeTypeMappersCaches[core.IfElse(index != -1, index, len(c.activeTypeMappersCaches)-1)]
+	if cachedType, ok := cache[key]; ok {
+		return cachedType
+	}
 	c.TotalInstantiationCount++
 	c.instantiationCount++
 	c.instantiationDepth++
 	result := c.instantiateTypeWorker(t, m, alias)
+	if index == -1 {
+		c.popActiveMapper()
+	} else {
+		cache[key] = result
+	}
 	c.instantiationDepth--
 	return result
 }
 
+var activeMapperCache = sync.Pool{
+	New: func() any {
+		return make(map[string]*Type, 1)
+	},
+}
+
+func (c *Checker) pushActiveMapper(mapper *TypeMapper) {
+	c.activeMappers = append(c.activeMappers, mapper)
+
+	c.activeTypeMappersCaches = append(c.activeTypeMappersCaches, activeMapperCache.Get().(map[string]*Type))
+}
+
+func (c *Checker) popActiveMapper() {
+	c.activeMappers[len(c.activeMappers)-1] = nil
+	c.activeMappers = c.activeMappers[:len(c.activeMappers)-1]
+
+	cache := c.activeTypeMappersCaches[len(c.activeTypeMappersCaches)-1]
+	c.activeTypeMappersCaches[len(c.activeTypeMappersCaches)-1] = nil
+	clear(cache)
```

Comment:
Something I'm not clear on - why do you need to clear the cache in addition to setting the old map to `nil`?

And why does the old code not just set them all to `undefined` instead of clearing all of them and keeping them in place?

Reply:
It's cleared before I `Put` it back in the pool.

Reply:
That being said, I should take an approach more like https://github.com/microsoft/TypeScript/pull/62016 and just keep the maps in the slice but empty.